rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,307 @@
+        self.current_fencing_token.lock().unwrap().clone()
+    }
+
+    async fn internal_acquire(
+        &self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+    ) -> Result<HybridLogicalClock, Error> {
+        let state_store_response = self
+            .state_store
+            .set(
+                self.lease_name.clone(),
+                self.lease_holder_name.clone(),
+                request_timeout,
+                None,
+                SetOptions {
+                    set_condition: SetCondition::OnlyIfEqualOrDoesNotExist,
+                    expires: Some(lease_expiration),
+                },
+            )
+            .await?;
+
+        if state_store_response.response {
+            self.current_fencing_token
+                .lock()
+                .unwrap()
+                .clone_from(&state_store_response.version);
+
+            state_store_response
+                .version
+                .ok_or(Error(ErrorKind::MissingFencingToken))
+        } else {
+            *self.current_fencing_token.lock().unwrap() = None;
+
+            Err(Error(ErrorKind::KeyAlreadyLeased))
+        }
+    }
+
+    /// Attempts to acquire a lease, returning if it cannot be acquired after one attempt.
+    ///
+    /// `lease_expiration` is how long the lease will remain held in the State Store after acquired, if not released before then.
+    /// `request_timeout` is the maximum time the function will wait for receiving a response from the State Store service, it is rounded up to the nearest second.
+    /// `renewal_period` is the frequency with which the lease will be auto-renewed by the lease client if acquired succesfully. `None` indicates the lease should not be auto-renewed.
+    ///
+    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or `Error` if lease is not acquired.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `Set` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// [`struct@Error`] of kind [`KeyAlreadyLeased`](ErrorKind::KeyAlreadyLeased) if the `lease` is already in use by another holder
+    ///
+    /// [`struct@Error`] of kind [`MissingFencingToken`](ErrorKind::MissingFencingToken) if the fencing token in the service response is empty.
+    ///
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
```

Comment:
This is changed to an error now, so update doc to remove the Panics section and add a doc for the new error (same on lock() doc)

Reply:
Gahh. ADHD is tough.